src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+----------------------------------------------------------------------------
+module Algebra.Graph.Bipartite.AdjacencyMap (
+    -- * Data structure
+    AdjacencyMap, leftAdjacencyMap, rightAdjacencyMap,
+
+    -- * Basic graph construction primitives
+    empty, leftVertex, rightVertex, vertex, edge, overlay, connect, vertices, edges,
+    overlays, connects,
+
+    -- * Conversion functions
+    toBipartite, fromBipartite, fromGraph,
+
+    -- * Graph properties
+    isEmpty, hasEdge, hasLeftVertex, hasRightVertex, hasVertex, leftVertexCount,
+    rightVertexCount, vertexCount, edgeCount,
+
+    -- * Miscellaneous
+    consistent,
+    ) where
+
+import Data.Either (lefts, rights)
+import Data.List   (sort)
+import Data.Tuple  (swap)
+import GHC.Generics
+
+import qualified Algebra.Graph              as G
+import qualified Algebra.Graph.AdjacencyMap as AM
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Set        as Set
+
+{-| The 'Bipartite.AdjacencyMap' data type represents an __undirected__
+bipartite graph by two maps of vertices into their adjacency sets. The two type
+parameteters define the types of identifiers of the vertices of each part.
+
+__Note:__ even if the identifiers and their types for two vertices of
+different parts are equal, these vertices are considered to be different.
+See examples for more details.
+-}
+data AdjacencyMap a b = BAM {
+    -- | The /adjacency map/ of the left part of the graph: each vertex is
+    -- associated with a set of its neighbours. Complexity: /O(1)/ time and
+    -- memory.
+    --
+    -- @
+    -- leftAdjacencyMap 'empty'                    == Map.'Map.empty'
+    -- leftAdjacencyMap ('leftVertex' 1)           == Map.'Map.singleton' 1 Set.'Set.empty'
+    -- leftAdjacencyMap ('rightVertex' 1)          == Map.'Map.empty'
+    -- leftAdjacencyMap ('edge' 1 2)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
+    -- leftAdjacencyMap ('edge' 1 1)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 1)
+    -- leftAdjacencyMap ('edges' [(1, 1), (1, 2)]) == Map.'Map.singleton' 1 (Set.'Set.fromAscList' [1, 2])
+    -- @
+    leftAdjacencyMap :: Map.Map a (Set.Set b),
+
+    -- | The inverse map for 'leftAdjacencyMap'. Complexity: /O(1)/ time and memory.
+    --
+    -- @
+    -- rightAdjacencyMap 'empty'                    == Map.'Map.empty'
+    -- rightAdjacencyMap ('leftVertex' 1)           == Map.'Map.empty'
+    -- rightAdjacencyMap ('rightVertex' 1)          == Map.'Map.singleton' 2 Set.'Set.empty'
+    -- rightAdjacencyMap ('edge' 1 2)               == Map.'Map.singleton' 2 (Set.'Set.singleton' 1)
+    -- rightAdjacencyMap ('edge' 1 1)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 1)
+    -- rightAdjacencyMap ('edges' [(1, 1), (1, 2)]) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+    -- @
+    rightAdjacencyMap :: Map.Map b (Set.Set a)
+} deriving (Eq, Show, Generic)
```

Comment:
You shouldn't use derived `Show` instance here. Please have a look how `Show` is implemented for `AdjacencyMap` and try to follow it here.

Reply:
Also, using derived `Eq` is probably suboptimal: we can implement `Eq` by comparing only one of the two maps.

Reply:
Oh wait, my second comment is probably wrong :)

Reply:
We can compare only `leftAdjacencyMap`s and `keysSet`s of the `rigthAdjacencyMap`s, this would be a little bit faster.